harambe/parser/parser.py

```diff
@@ -159,6 +171,79 @@ def _schema_to_pydantic_model(
 
         return create_model(model_name, __config__=config, **fields)  # type: ignore
 
+    def _get_all_required_fields(
+        self, schema: Schema, parent_key: str = ""
+    ) -> List[str]:
+        """
+        Recursively collect all required fields from the schema, including nested ones.
+        """
+        required_fields = []
+        for field_name, field_info in schema.items():
+            full_key = f"{parent_key}.{field_name}" if parent_key else field_name
+            if field_info.get("required", False):
+                required_fields.append(full_key)
+
+            if field_info.get("type") == OBJECT_TYPE:
+                nested_properties = field_info.get("properties", {})
+                required_fields.extend(
+                    self._get_all_required_fields(nested_properties, full_key)
+                )
+            elif field_info.get("type") == LIST_TYPE:
+                item_info = field_info.get("items", {})
+                if item_info.get("type") == OBJECT_TYPE:
+                    nested_properties = item_info.get("properties", {})
+                    required_fields.extend(
+                        self._get_all_required_fields(nested_properties, full_key)
+                    )
+
+        return required_fields
+
+    def _find_missing_required_fields(
+        self, data: dict[str, Any], required_fields: List[str]
+    ) -> List[str]:
+        """
+        Check the data against the list of required fields and find which are missing or None.
+        """
+        missing_fields = []
+
+        def is_empty(value: Any) -> bool:
+            if value is None:
+                return True
+            if isinstance(value, str) and not value.strip():
+                return True
+            if isinstance(value, (list, dict)) and not value:
+                return True
+            return False
+
+        def check_value(data: Any, field_path: str):
+            keys = field_path.split(".")
+            for key in keys[:-1]:
+                if isinstance(data, dict):
+                    data = data.get(key, None)
+                elif isinstance(data, list):
+                    data = [
+                        item.get(key, None) if isinstance(item, dict) else None
+                        for item in data
+                    ]
+                else:
+                    return
+
+            final_key = keys[-1]
+            if isinstance(data, dict):
+                if is_empty(data.get(final_key)):
+                    missing_fields.append(field_path)
+            elif isinstance(data, list):
+                for item in data:
+                    if isinstance(item, dict) and is_empty(item.get(final_key)):
+                        missing_fields.append(field_path)
+            elif final_key and is_empty(data):
+                missing_fields.append(field_path)
+
+        for field in required_fields:
```

Comment:
i think we could make this simpler by iterating the output. then for each key we check if it is required. Break and raise and error on the first occurence. We dont need to find them all

Reply:
We should find them all actually. Best to get the entire error context to avoid both humans and agents from having to get error -> iterate -> get another error -> iterate -> get another error -> etc